Room from rt60

fakufaku opened the pull request:
Adds an option to `pyroomacoustics.room.ShoeBox` so that the reverberation time can be specified. The absorption and max_order are infered automatically using Sabine's formula.

This PR addresses issue #18

fakufaku commented:
Ideally, I would like to add some unit testing, but I'm not sure the 3D case gives very good results now. For `rt60=0.3` it is good, but as the longer the `rt60` the more off it gets.

ebezzam commented on examples/room_rt60.py:
```diff
@@ -0,0 +1,101 @@
+        # show result
+        plt.plot(T, power)
+        plt.plot(T, a * T + b, '--')
+        plt.plot(T, np.ones_like(T) * -60, '--')
```
I'm wondering if here's better to add the offset to the `-60` line? Now it's like this:

![original](https://user-images.githubusercontent.com/4757445/43777432-18578d4e-9a53-11e8-9492-6718bf0e7e2a.png)

But I'm thinking something like this is more representative of a `-60 dB` **drop** rather than `-60 dB`.

![offset](https://user-images.githubusercontent.com/4757445/43777492-41065b76-9a53-11e8-82e3-4fa427c8eec2.png)

ebezzam replied:
with the offset that you computed for the orange line

fakufaku replied:
Hum, this is an annoying issue. It stems from not having a proper method for computing the empirical T60. I think either is fine now, so I'll go with the more correct version you propose. The correct method will come in a subsequent pull request.

ebezzam commented on pyroomacoustics/room.py:
```diff
@@ -1164,4 +1204,55 @@ def extrude(self, height):
+        # surface computation is diff for 2D and 3D
+        if len(room_dim) == 2:
+            S = np.sum(room_dim)
+            sab_coef = 5  # The Sabine's coefficient needs to be adjusted in 2D
```
For the 2D case, I've trying to get an intuitive understanding of what these coefficients/constrants should be. Here's what I've been able to dig up so far.

Video tutorial (skipping to the constant part): https://www.youtube.com/watch?v=A8tWMnr5pps&feature=youtu.be&t=11m22s

Written up derivation of Sabine: https://courses.physics.illinois.edu/phys406/sp2017/Lecture_Notes/P406POM_Lecture_Notes/Derivation_of_the_Sabine_Equation.pdf

So in the well-known [3D case](https://en.wikipedia.org/wiki/Reverberation#Sabine_equation) there's this constant factor of `24` which seems to come from a factor `4` which comes computing the surface area of the sphere multiplied by a factor `6` dropping down from `ln(10^-6)` for the `-60` drop.

So for 2D, I think the combined factor should be `2*6=12` where `2` comes from the area of a circle and the `6` still comes from the `-60 dB` drop.

fakufaku replied:
Amazing! Thanks for doing the digging! That falls square in the range that seems correct. Ampirically, I had this 5, but after correcting the computation of the perimeter, this should become 10. That is quite close to the correct 12 that you derived. I'll make the change.

ebezzam commented on pyroomacoustics/room.py:
```diff
@@ -1164,4 +1204,55 @@ def extrude(self, height):
+        V = np.prod(room_dim)  # volume
+        # surface computation is diff for 2D and 3D
+        if len(room_dim) == 2:
+            S = np.sum(room_dim)
```
Here I think it should be `2*np.sum(room_dim)` to sum the 2D room boundaries. This adds up just two of the walls.

fakufaku replied:
Right!

fakufaku commented:
@ebezzam I have implemented a routine to measure the RT60 automatically in `pyroomacoustics.experimental.measure_rt60`. I have also corrected the computation of absorbtion factor according to your feedback. However, the generated IR and the target RT60 do not match. Especially when the reverberation time is long. This is a bit strange. Maybe we are missing something somewhere ?

fakufaku commented:
Somewhat resolved by the addition of [inverse Sabine function](https://pyroomacoustics.readthedocs.io/en/pypi-release/pyroomacoustics.acoustics.html#pyroomacoustics.acoustics.inverse_sabine) in PR #75 .
